torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+    return node_type == other.node_type && key_size == other.key_size &&
+        std::memcmp(key, other.key, key_size) == 0;
+  }
+
+  size_t hash() const {
+    // don't bother hashing the key data, common case 1 cache entry per node
+    return std::hash<std::type_index>()(node_type) ^ key_size;
+  }
+
+  std::type_index node_type;
+  uint16_t key_size;
+  const uint8_t* key;
+};
+
+struct OutputRef {
+  OutputRef() : node_id(-1), index(-1) {}
+  OutputRef(int node_, int index_) : node_id(node_), index(index_) {}
+
+  bool is_set() const {
+    return this->node_id >= 0;
+  }
+
+  int node_id;
+  int index;
+};
+
+struct NodeCall {
+  NodeCall(const std::shared_ptr<Node>& node_)
+      : node(node_), input_refs(node_->num_inputs()) {}
+
+  NodeCall(const std::shared_ptr<Node>& node_, int num_inputs_)
+      : node(node_), input_refs(num_inputs_) {}
+
+  OutputRef& operator[](size_t pos) {
+    return input_refs[pos];
+  }
+
+  size_t size() const {
+    return input_refs.size();
+  }
+
+  void mark_output(int input_nr, int output_idx) {
+    graph_output.emplace_back(std::make_pair(input_nr, output_idx));
+  }
+
+  std::shared_ptr<Node> node;
+  // at::SmallVector??
+  std::vector<OutputRef> input_refs;
+
+  // borrowed references
+  std::vector<std::pair<int, int>> tensor_pre_hooks;
+  std::vector<int> pre_hooks;
+  std::vector<int> post_hooks;
+  std::vector<std::pair<int, int>> graph_output;
+};
+
+struct AutogradCompilerCall {
+  AutogradCompilerCall(bool accumulate_grad_)
+      : accumulate_grad(accumulate_grad_),
+        default_dyn_type(SizeInput::STATIC) {}
+
+  void add_tensor_input(const at::Tensor& tensor) {
+    inputs.emplace_back(tensor);
+  }
+
+  void add_set_grad_target(const at::Tensor& tensor) {
+    set_grad_targets.emplace_back(tensor);
+  }
+
+  void add_size_input(const c10::SymInt& s) {
+    all_size_inputs.emplace_back(SizeInput(default_dyn_type, s.expect_int()));
+  }
+
+  int emplace_hook(c10::SafePyObject&& fn) {
+    hooks.emplace_back(std::move(fn));
+    return hooks.size() - 1;
+  }
+
+  std::vector<SizeInput> all_size_inputs;
+  std::vector<int64_t> dyn_size_inputs;
+  std::vector<at::Tensor> inputs;
+  std::vector<at::Tensor> set_grad_targets;
+  std::vector<c10::SafePyObject> hooks;
+  bool accumulate_grad;
+  SizeInput::DynType default_dyn_type;
+};
+
+class CompiledNodeArgs {
+  // CompiledNodeArgs builds a representation of the constant values found
+  // across all the nodes in the compiled graph, via 'collect' overloads. The
+  // collected constants are specialized on by concatenation into a cache key.
+  // Tensor, symint arguments (which are lifted to become graph inputs rather
+  // than specialized on) are forwarded to the compiler and not included in the
+  // key.
+ public:
+  void collect(const at::Tensor& t) {
+    collect(t.defined());
+    if (t.defined()) {
+      _compiler.add_tensor_input(t);
+    }
+  }
+  void collect(const SavedVariable& t) {
+    collect(t.unpack(_node_call.node));
```

Comment:
@albanD is it OK to unpack saved variables multiple times?

Reply:
The unpack hooks will trigger multiple times.
In general it is not ok no, things like checkpointing (or the Simple(r) FSDP impl) will be very unhappy about this.

Reply:
Is there a way to peek into a SavedVariable without triggering side effects?

Reply:
What do you want to get from it. When hooks are involved, the SavedVariable might contain anything.
When doing AC for example, calling unpack will trigger a recompute of the whole forward pass to re-generate this Tensor.

Reply:
We also need to specialize on if this is defined.

Maybe this is a new type of hook that needs to be lifted to an input.

Reply:
I added an assert there is no callback.